Automattic/jetpack
# Jetpack: optimize the way we initialize ToS

## Proposed changes:
* Optimize the way we initialize ToS component.

Tracking of Jetpack-the-plugin events is initialized on-demand via `do_action( 'jetpack_initialize_tracking' )`.
The action will first check if tracking is allowed to be initialized.

### Other information:


## Jetpack product discussion
pdWQjU-QN-p2#comment-918
https://github.com/Automattic/vulcan/issues/503

## Does this pull request change what data or activity we track or use?
We modify the way we initialize tracking, but do not add or remove any tracking events.

## Testing instructions:
1. Disconnect Jetpack if connected, but set the `jetpack_tos_agreed` option (`wp option update jetpack_tos_agreed 1`).
2. Edit the file`/jetpack/src/class-tracking.php`, find the `init()` method and add the following code before `static::$initialized = true;`:
```php
error_log( 'Jetpack Tracking initialized' );
```
3. Visit a few non-Jetpack admin pages, confirm `debug.log` stays empty.
4. Visit "My Jetpack", confirm `Jetpack Tracking initialized` line shows up in `debug.log` (there'll be a few due to API requests).
5. Visit other Jetpack pages, confirm new log entries show up for each.
6. Run an XML-RPC request, confirm new log entry gets added:
```bash
curl -is -H 'Content-Type: text/xml' --data '<?xml version="1.0"?><methodCall><methodName>system.listMethods</methodName><params></params></methodCall>' 'https://yours-site.example/xmlrpc.php?for=jetpack' && echo
```
7. Open a REST route in the browser (e.g. `https://your-site.example/?rest_route=%2Fjetpack%2Fv4%2Fconnection`), confirm new log entry gets added.
8. Connect Jetpack, confirm new entries show up even on non-Jetpack pages.
